#!/usr/bin/env cwl-runner

$namespaces:
  dct: http://purl.org/dc/terms/
  foaf: http://xmlns.com/foaf/0.1/
  doap: http://usefulinc.com/ns/doap#
  adms: http://www.w3.org/ns/adms#
  dcat: http://www.w3.org/ns/dcat#

$schemas:
- http://dublincore.org/2012/06/14/dcterms.rdf
- http://xmlns.com/foaf/spec/20140114.rdf
- http://usefulinc.com/ns/doap#
- http://www.w3.org/ns/adms#
- http://www.w3.org/ns/dcat.rdf

cwlVersion: "cwl:draft-3.dev3"

class: CommandLineTool

adms:includedAsset:
  doap:name: "UCSC userApps"
  doap:description: |
    UCSC genome browser 'kent' bioinformatic utilities
    These are only the command line bioinformatic utilities
    from the kent source tree.
  doap:homepage: "http://genome.ucsc.edu/goldenPath/help/bigWig.html"
  dcat:downloadURL: "http://hgdownload.cse.ucsc.edu/admin/exe/userApps.v325.src.tgz"
  doap:release:
  - class: doap:Version
    doap:revision: "v325"
  doap:license: "GPL"
  doap:category: "commandline tool"
  doap:programming-language: "C"
  foaf:publications:
  - id: urn:pmid:20639541
    foaf:title: "(Kent et al., 2010) BigWig and BigBed: enabling browsing of large distributed datasets. Bioinformatics."
    foaf:homepage: "http://www.ncbi.nlm.nih.gov/pubmed/20639541"
  doap:developer:
  - class: foaf:Organization
    foaf:name: "CIRM Stem Cell Genomics Data Management Center"
    foaf:member:
    - class: foaf:Person
      foaf:name: "Jim Kent"
      foaf:mbox: "mailto:kent@soe.ucsc.edu"

description: |
  ucsc-bedGraphToBigWig.cwl is developed for CWL consortium

  Original tool usage:
         bedGraphToBigWig in.bedGraph chrom.sizes out.bw

      where in.bedGraph is a four column file in the format:
            <chrom> <start> <end> <value>
      and chrom.sizes is a two-column file/URL: <chromosome name> <size in bases>
      and out.bw is the output indexed big wig file.
      If the assembly <db> is hosted by UCSC, chrom.sizes can be a URL like
        http://hgdownload.cse.ucsc.edu/goldenPath/<db>/bigZips/<db>.chrom.sizes
      or you may use the script fetchChromSizes to download the chrom.sizes file.
      If not hosted by UCSC, a chrom.sizes file can be generated by running
      twoBitInfo on the assembly .2bit file.
      The input bedGraph file must be sorted, use the unix sort command:
        sort -k1,1 -k2,2n unsorted.bedGraph > sorted.bedGraph

doap:name: "ucsc-bedGraphToBigWig.cwl"
dcat:downloadURL: "https://github.com/common-workflow-language/workflows/blob/master/tools/ucsc-bedGraphToBigWig.cwl"

doap:maintainer:
- class: foaf:Organization
  foaf:name: "Barski Lab, Cincinnati Children's Hospital Medical Center"
  foaf:member:
  - class: foaf:Person
    id: "http://orcid.org/0000-0001-9102-5681"
    foaf:openid: "http://orcid.org/0000-0001-9102-5681"
    foaf:name: "Andrey Kartashov"
    foaf:mbox: "mailto:Andrey.Kartashov@cchmc.org"

requirements:
  - class: InlineJavascriptRequirement
  - $import: envvar-global.cwl
  - $import: ucsc-userapps-docker.cwl

inputs:
  - id: "#input"
    type: File
    inputBinding:
      position: 2

  - id: "#genomeFile"
    type: File
    inputBinding:
      position: 3

  - id: "#bigWig"
    type: string
    inputBinding:
      position: 4

  - id: "#unc"
    type: ["null",boolean]
    description: "If set, do not use compression."
    inputBinding:
      position: 1
      prefix: "-unc"

  - id: "#itemsPerSlot"
    type: ["null",int]
    description: |
      -itemsPerSlot=N - Number of data points bundled at lowest level. Default 1024
    inputBinding:
      separate: false
      position: 1
      prefix: "-itemsPerSlot="

  - id: "#blockSize"
    type: ["null",int]
    description: |
      -blockSize=N - Number of items to bundle in r-tree.  Default 256
    inputBinding:
      separate: false
      position: 1
      prefix: "-blockSize="

outputs:
  - id: "#bigWigOut"
    type: File
    outputBinding:
      glob: $(inputs.bigWig)

baseCommand: ["bedGraphToBigWig"]
